File: 124.Binary Tree Maximum Path Sum.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var maxPathSum = function (root) {
    update(root);
    return root.max;
};

var update = function (root) {
    var t = root;
    if (t === null) return;
    update(t.left);
    update(t.right);
    if (t.left === null && t.right === null) {
        t.singleMax = t.val;
        t.doubleMax = t.val;
        t.max = t.val;
    } else if (t.left === null) {
        t.singleMax = t.val + t.right.singleMax > t.val ? t.val + t.right.singleMax : t.val;
        t.doubleMax = t.val + t.right.singleMax > t.val ? t.val + t.right.singleMax : t.val;
        t.max = t.doubleMax > t.right.max ? t.doubleMax : t.right.max;
    } else if (t.right === null) {
        t.singleMax = t.val + t.left.singleMax > t.val ? t.val + t.left.singleMax : t.val;
        t.doubleMax = t.val + t.left.singleMax > t.val ? t.val + t.left.singleMax : t.val;
        t.max = t.doubleMax > t.left.max ? t.doubleMax : t.left.max;
    } else {
        t.doubleMax = t.val + (t.left.singleMax > 0 ? t.left.singleMax : 0) + (t.right.singleMax > 0 ? t.right.singleMax : 0);
        tmp = t.left.singleMax > t.right.singleMax ? t.left.singleMax : t.right.singleMax;
        t.singleMax = t.val + (tmp > 0 ? tmp : 0);
        t.max = t.doubleMax > t.left.max ? t.doubleMax : t.left.max;
        t.max = t.max > t.right.max ? t.max : t.right.max;
    }
}

File: 124.Binary Tree Maximum Path Sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
private:
  int ans = numeric_limits<int>::min();
  int maxPathDownwards(TreeNode* root) {
    if (!root) return 0;
    int left = maxPathDownwards(root->left);
    int right = maxPathDownwards(root->right);
    ans = max(ans, (max(0, left) + max(0, right) + root->val));
    return root->val + max({0, left, right});
  }
public:
  int maxPathSum(TreeNode* root) {
    maxPathDownwards(root);
    return ans;
  }
};
